### PR TITLE
Move import source management from Model to Importer.

### LINKED ISSUE #802: Remove ImportSource management from Model class.
I would like to see the ImportSource management removed from the Model class.
I think management of ImportSources should be handled by the Importer class.

### PR DESCRIPTION
**Pull request by hsorby**
Closes #802.

### CODE REVIEW

File: src/api/libcellml/model.h [hunk 1]
```diff
@@ -343,98 +343,6 @@ class LIBCELLML_EXPORT Model: public ComponentEntity
...
-    /**
-     * @brief Remove any empty components and units from the model.
-     * 
-     *  Remove any empty components and units fom this model.
-     *  In this context, "empty" means:
-     *   - components with no name, id, resets, variables, maths, or non-empty child components; and
-     *   - units which have no name, id, or child units.
-     *  For components, this is applied recursively from the leaves of the encapsulation hierarchy.
-     */
-    void clean();
```
Comment (nickerso) [inline src/api/libcellml/model.h]: This needs to go back in :)

File: src/bindings/interface/model.i [hunk 2]
```diff
@@ -66,28 +66,6 @@ Only the first matching Units is removed and returned.";
...
-"Remove an ImportSource from the Model by index or reference.";
-
-%feature("docstring") libcellml::Model::removeAllImportSources
-"Remove all ImportSource items from the Model.";
-
-%feature("docstring") libcellml::Model::hasImportSource
-"Test whether this Model already contains the given ImportSource.";
-
-%feature("docstring") libcellml::Model::clean
-"Remove any empty units and any empty components from the model."
```
Comment (nickerso) [inline src/bindings/interface/model.i]: go back in

File: src/printer.cpp [hunk 3]
```diff
@@ -397,16 +397,46 @@ std::string Printer::PrinterImpl::printReset(const ResetPtr &reset, IdList &idLi
...
+//    if (!importedEntity->id().empty()) {
+//        repr += " id=\"" + importedEntity->id() + "\"";
+//    } else if (autoIds) {
+//        repr += " id=\"" + makeUniqueId(idList) + "\"";
+//    }
+//    repr += "/>";
+
+//    return repr;
+//}
+
```
Comment (nickerso) [inline src/printer.cpp]: should all be removed?

File: tests/bindings/python/test_model.py [hunk 4]
```diff
@@ -350,73 +350,6 @@ def test_variable_interfaces(self):
...
-        self.assertEqual(4, model.componentCount())
-        self.assertEqual(3, model.unitsCount())
-
-        # Call the Model.clean() function to remove the empty items (c4 and u3).
-        model.clean()
-
-        self.assertEqual(3, model.componentCount())
-        self.assertEqual(2, model.unitsCount())
-
-
```
Comment (nickerso) [inline tests/bindings/python/test_model.py]: clean test needs to go back in

File: tests/import_source/import_source.cpp [hunk 5]
```diff
@@ -28,128 +28,128 @@ TEST(ImportSource, createImportSource)
...
 TEST(ImportSource, addImportSourceBadInput)
 {
-    auto model = libcellml::Model::create();
+    auto importer = libcellml::Importer::create();
 
-    EXPECT_FALSE(model->addImportSource(nullptr));
+    EXPECT_FALSE(importer->addImportSource(nullptr));
 }
 
 TEST(ImportSource, addToModel)
```
Comment (nickerso) [inline tests/import_source/import_source.cpp]: rename test `addToImporter` ?

File: tests/import_source/import_source.cpp [hunk 6]
```diff
@@ -177,38 +177,38 @@ TEST(ImportSource, importSourceMove)
 
 TEST(ImportSource, importSourceSharedAcrossModels)
```
Comment (nickerso) [inline tests/import_source/import_source.cpp]: rename test

File: tests/import_source/import_source.cpp [hunk 7]
```diff
@@ -177,38 +177,38 @@ TEST(ImportSource, importSourceMove)
...
 {
-    auto m1 = libcellml::Model::create("m1");
+    auto i1 = libcellml::Importer::create();
     auto imp = libcellml::ImportSource::create();
 
-    EXPECT_TRUE(m1->addImportSource(imp));
+    EXPECT_TRUE(i1->addImportSource(imp));
 
     // Add import source to another model.
```
Comment (nickerso) [inline tests/import_source/import_source.cpp]: Add import source to another IMPORTER. and similar comments throughout...

File: tests/import_source/import_source.cpp (hunk 6 above)
Comment (nickerso) [reply tests/import_source/import_source.cpp]: and many others test names...

**Review by nickerso — CHANGES_REQUESTED** [review]
A few minor renaming that has been missed in comments and test names, and also seems to have pulled some of the Model::clean() work that has gone in recently.

File: tests/import_source/import_source.cpp (hunk 7 above)
Comment (hsorby) [reply tests/import_source/import_source.cpp]: 2dec745

File: tests/import_source/import_source.cpp (hunk 6 above)
Comment (hsorby) [reply tests/import_source/import_source.cpp]: 2dec745

File: tests/import_source/import_source.cpp (hunk 5 above)
Comment (hsorby) [reply tests/import_source/import_source.cpp]: 2dec745

File: src/printer.cpp (hunk 3 above)
Comment (hsorby) [reply src/printer.cpp]: b49610c

File: tests/bindings/python/test_model.py (hunk 4 above)
Comment (hsorby) [reply tests/bindings/python/test_model.py]: eaa552f

File: src/bindings/interface/model.i (hunk 2 above)
Comment (hsorby) [reply src/bindings/interface/model.i]: eaa552f

File: src/api/libcellml/model.h (hunk 1 above)
Comment (hsorby) [reply src/api/libcellml/model.h]: eaa552f

**Review by nickerso — DISMISSED** [review]

File: src/annotator.cpp [hunk 8]
```diff
@@ -912,14 +912,15 @@ void Annotator::clearAllIds()
...
-            model->units(i)->setId("");
-            for (size_t j = 0; j < model->units(i)->unitCount(); ++j) {
-                model->units(i)->setUnitId(j, "");
+            auto units = model->units(i);
+            units->removeId();
+            if (units->isImport()) {
+                units->importSource()->removeId();
+            }
+            for (size_t j = 0; j < units->unitCount(); ++j) {
+                units->setUnitId(j, "");
```
Comment (agarny) [inline src/annotator.cpp:927]: If we have `removeId()` then surely we should have `removeUnitId()`?

File: src/annotator.cpp [hunk 9]
```diff
@@ -943,18 +944,23 @@ void Annotator::clearAllIds(ModelPtr &model)
...
     for (size_t i = 0; i < component->variableCount(); ++i) {
-        component->variable(i)->setId("");
-        for (size_t j = 0; j < component->variable(i)->equivalentVariableCount(); ++j) {
-            Variable::setEquivalenceConnectionId(component->variable(i), component->variable(i)->equivalentVariable(j), "");
-            Variable::setEquivalenceMappingId(component->variable(i), component->variable(i)->equivalentVariable(j), "");
+        auto variable = component->variable(i);
+        variable->removeId();
+        for (size_t j = 0; j < variable->equivalentVariableCount(); ++j) {
+            Variable::setEquivalenceConnectionId(variable, variable->equivalentVariable(j), "");
+            Variable::setEquivalenceMappingId(variable, variable->equivalentVariable(j), "");
```
Comment (agarny) [inline src/annotator.cpp:960]: Should have a `removeXxxxId()`.

File: src/annotator.cpp [hunk 10]
```diff
@@ -1714,25 +1720,27 @@ size_t Annotator::AnnotatorImpl::generateHash()
...
+        size_t i;
         idsString += "m=" + model->id() + "me=" + model->encapsulationId();
 
-        for (size_t u = 0; u < model->importSourceCount(); ++u) {
-            auto import = model->importSource(u);
-            idsString += "i=" + std::to_string(u) + import->id();
+        auto importSources = getAllImportSources(model);
+        i = 0;
+        for (auto &importSource : importSources) {
+            idsString += "i=" + std::to_string(i++) + importSource->id();
```
Comment (agarny) [inline src/annotator.cpp]: Would replace with:
```c++
i = MAX_SIZE_T;
for (auto &importSource : importSources) {
    idsString += "i=" + std::to_string(++i) + importSource->id();
```

**Review by agarny — CHANGES_REQUESTED** [review]
Just some minor comments.

File: src/annotator.cpp (hunk 8 above)
Comment (hsorby) [reply src/annotator.cpp:927]: Yes, but not in this PR.

File: src/annotator.cpp (hunk 9 above)
Comment (hsorby) [reply src/annotator.cpp:960]: Yes, but not in this PR.

File: src/annotator.cpp (hunk 10 above)
Comment (hsorby) [reply src/annotator.cpp]: ba2e9e1

**Review by agarny — APPROVED** [review]

**Review by nickerso — APPROVED** [review]

**Pull request merged**